src/VisualStudio/Core/Def/Options/VisualStudioOptionPersisterProvider.cs

```diff
@@ -50,11 +49,27 @@ internal sealed class VisualStudioOptionPersisterProvider : IOptionPersisterProv
         }
 
         public async ValueTask<IOptionPersister> GetOrCreatePersisterAsync(CancellationToken cancellationToken)
-            => _lazyPersister ??=
-                new VisualStudioOptionPersister(
-                    new VisualStudioSettingsOptionPersister(RefreshOption, _readFallbacks, await _settingsManagerService.GetValueAsync(cancellationToken).ConfigureAwait(true)),
-                    await LocalUserRegistryOptionPersister.CreateAsync(_localRegistryService).ConfigureAwait(false),
-                    new FeatureFlagPersister(await TryGetServiceAsync(_featureFlagsService).ConfigureAwait(false)));
+        {
+            return _lazyPersister ??= await CreatePersisterAsync(cancellationToken).ConfigureAwait(true);
+
+            async Task<VisualStudioOptionPersister> CreatePersisterAsync(CancellationToken cancellationToken)
+            {
+                // Obtain services before creating instances. This avoids state corruption in the event cancellation is
+                // requested (some of the constructors register event handlers that could leak if cancellation occurred
+                // in the middle of construction).
+                var settingsManager = await _settingsManagerService.GetValueAsync(cancellationToken).ConfigureAwait(true);
```

Comment:
Should we just not have cancellation here? I can't imagine we're really passing a meaningful cancellation token here.

Reply:
➡️ It can happen if VS closes shortly after launch